File: julia/roman-numerals/roman-numerals.jl
function to_roman(number::Integer)::String
    map = (
        "M" => 1000,
        "CM" => 900,
        "D" => 500,
        "CD" => 400,
        "C" => 100,
        "XC" => 90,
        "L" => 50,
        "XL" => 40,
        "X" => 10,
        "IX" => 9,
        "V" => 5,
        "IV" => 4,
        "I" => 1
    )

    if number <= 0
        throw(error("number must be positive"))
    end
    result = ""
    for (roman, int) in map
        count = div(number, int)
        result *= repeat(roman, count)
        number %= int
    end

    return result
end

File: solutions/go/roman-numerals/1/roman_numerals.go
package romannumerals

import (
	"errors"
	"strings"
)

type romanMapping struct {
	value  int
	symbol string
}

var romanMappings = []romanMapping{
	{1000, "M"},
	{900, "CM"},
	{500, "D"},
	{400, "CD"},
	{100, "C"},
	{90, "XC"},
	{50, "L"},
	{40, "XL"},
	{10, "X"},
	{9, "IX"},
	{5, "V"},
	{4, "IV"},
	{1, "I"},
}

func ToRomanNumeral(input int) (string, error) {
	if input < 1 || input > 3999 {
		return "", errors.New("input must be between 1 and 3999")
	}

	result := ""
	for _, mapping := range romanMappings {
		count := input / mapping.value
		result += strings.Repeat(mapping.symbol, count)
		input = input % mapping.value
	}
	return result, nil
}
